[DNM] WIP

yuzefovich opened the pull request.

yuzefovich committed:
colexec: extract new colexecutils package

This commit extracts new `colexecutils` package from `colexec`, and the
new package now contains the following:
- the code for the spilling queue (which has been exported)
- `AppendOnlyBufferedBatch` wrapper (also exported) and
`MakeWindowIntoBatch` utility method.

Additionally, this commit moves `FeedOperator` and `NonExplainable` marker
into the `colexecbase` package and extracts the utility method to generate
random batch size for tests from `colexec` into `colexectestutils`.

Release note: None

yuzefovich committed:
colexec: extract new colexechash package

This commit extracts new `colexechash` package that currently contains
all of the code for the hash table as well as some hash utilities (like
tuple hash distributor). In order to get here the following changes have
been made:
- many things about the hash table have been exported
- `TupleHashDistributor` has been exported
- `CancelChecker` and some utility methods have been moved to
`colexecutils` package.

The reason for extracting this package is that I envision that we will
have the operators that use the hash table (hash joiner, hash
aggregator, unordered distinct) will likely live in separate packages,
so having separate "hash utility" package seems like a good idea.

Release note: None

yuzefovich committed:
colflow: move hash router into colflow package

This commit moves the hash router into `colflow` package since it is
a better place for it. This required some changes to the test in order
to instantiate the objects previously taken from `colexec/main_test.go`.
Additional complication is that `colflow/main_test.go` lives in
`colflow_test` package in order to avoid an import cycle, so we needed
to create some utility objects manually in `routers_test.go`.

Release note: None

yuzefovich committed:
colexec: break up operator.go file

This commits breaks up `operator.go` file in the following manner:
- move `OperatorInitStatus` into `colexecbase`
- export and move `CloserHelper` and `ClosableOperator` into
`colexecbase`
- move `twoInputNode` into newly created `colexec/joiner_utils.go` since
it is only used by two joiner operators
- export and move `OneInputCloserHelper` into `colexecbase`
- move `zeroOperator` and `fixedNumTuplesNoInputOp` into `colexecutils`
- move `vectorTypeEnforcer` and `BatchSchemaPrefixEnforcer` into
`colexecutils`.

Release note: None

yuzefovich committed:
colexec: extract new colexecwindow package

This commit extracts new `colexecwindow` package that currently contains
all of the code related to the window functions (several operators moved
from `colexec` as well as some utility information moved from
`colbuilder`). It is worth noting that at this moment this new package
depends on `colexec` for a few reasons (planning a Const operator,
a chain of ordered distinct operators, etc), but I intend to break up
that dependency later.

Release note: None

yuzefovich committed:
colexec: clean up like_ops code generation

This commit cleans up the generation of the code for `like_ops`.
Previously, the template depended on two other template files during
code generation which became quite tricky to work around with bazel.
Another complication is that we want to move the projection and the
selection operators into separate packages, and this like_ops template
was too tightly coupled.

This commit fixes the situation by breaking up the resulting generated
code to be in two separate files (so that we can later customize the
package name), but also this way we only depend on a single target which
cleans up the bazel build.

As another minor fix this commit also includes the text of CCL license
into the generated code for like operators.

Release note: None

yuzefovich committed:
colexec: extract op_creation into new colexecargs package

This commit extracts several structs that encompass the arguments to
and the result of `NewColOperator` call into new `colexecargs` package.

Additionally, `colexec/proj_utils_test.go` has been moved into
`colexectestutils` since in the future the utilities from the file will
be used from different packages.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60851)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: break up distinct ops and sort partitioners

`distinct_tmpl` is a bit special because it is used to generate code for
two different purposes: distinct structs (used by the ordered distinct)
and partitioner structs (used by the sort operators). Previously, all of
this would get generated into the same file making it harder to break
down `colexec` package, so this commit updates the corresponding code
generator to write those two things into two separate files.

Release note: None

yuzefovich committed:
colexec: extract new colexecdistinct package

This commit extracts new `colexecdistinct` package that is intended to
contain only the code related to the ordered distinct since that code is
used in several other places. This also required the following changes:
- `BoolVecToSelOp` is exported and moved into `colexecutils`
- some changes to the bazel build of `colexec` were needed in order to
generate `colexec/sort_partitioner.eg.go` that depends on `distinct_tmpl`
now residing in `colexecdistinct`.

Note that the unit tests for the ordered distinct are still kept in
`colexec` package since they are tightly coupled with the unordered and
external distinct unit tests.

Release note: None
